src/server/pkg/storage/chunk/writer.go

```diff
@@ -15,83 +15,73 @@ import (
 // Chunk split points are determined by a bit pattern in a rolling hash function (buzhash64 at https://github.com/chmduquesne/rollinghash).
 type Writer struct {
-	ctx        context.Context
-	objC       obj.Client
-	cbs        []func([]*DataRef) error
-	buf        *bytes.Buffer
-	hash       *buzhash64.Buzhash64
-	splitMask  uint64
-	dataRefs   []*DataRef
-	done       [][]*DataRef
-	rangeSize  int64
-	rangeCount int64
+	ctx            context.Context
+	objC           obj.Client
+	buf            *bytes.Buffer
+	hash           *buzhash64.Buzhash64
+	splitMask      uint64
+	f              WriterFunc
+	annotations    []*Annotation
+	chunkCount     int64
+	annotationSize int64
 }
 
 // newWriter creates a new Writer.
-func newWriter(ctx context.Context, objC obj.Client) *Writer {
+func newWriter(ctx context.Context, objC obj.Client, averageBits int, f WriterFunc) *Writer {
 	// Initialize buzhash64 with WindowSize window.
 	hash := buzhash64.New()
-	hash.Write(initialWindow)
-	return &Writer{
+	w := &Writer{
 		ctx:       ctx,
 		objC:      objC,
-		cbs:       []func([]*DataRef) error{},
 		buf:       &bytes.Buffer{},
 		hash:      hash,
-		splitMask: (1 << uint64(AverageBits)) - 1,
+		splitMask: (1 << uint64(averageBits)) - 1,
+		f:         f,
 	}
+	w.resetHash()
+	return w
 }
 
 func (w *Writer) resetHash() {
 	w.hash.Reset()
 	w.hash.Write(initialWindow)
 }
 
-// StartRange specifies the start of a range within the byte stream that is meaningful to the caller.
-// When this range has ended (by calling StartRange again or Close) and all of the necessary chunks are written, the
-// callback given during initialization will be called with DataRefs that can be used for accessing that range.
-func (w *Writer) StartRange(cb func([]*DataRef) error) {
-	// Finish prior range.
-	if w.dataRefs != nil {
-		w.finishRange()
+// Annotate associates an annotation with the next set of bytes that are written.
+func (w *Writer) Annotate(a *Annotation) {
+	a.Offset = int64(w.buf.Len())
+	// Handle the edge case when the last annotation from the prior chunk
+	// should not carry over into the next.
+	if a.Offset == 0 {
+		w.annotations = nil
 	}
-	// Start new range.
-	w.cbs = append(w.cbs, cb)
-	w.dataRefs = []*DataRef{&DataRef{OffsetBytes: int64(w.buf.Len())}}
-	w.rangeSize = 0
-	w.rangeCount++
-}
-
-func (w *Writer) finishRange() {
-	lastDataRef := w.dataRefs[len(w.dataRefs)-1]
-	lastDataRef.SizeBytes = int64(w.buf.Len()) - lastDataRef.OffsetBytes
-	data := w.buf.Bytes()[lastDataRef.OffsetBytes:w.buf.Len()]
-	lastDataRef.Hash = hash.EncodeHash(hash.Sum(data))
-	w.done = append(w.done, w.dataRefs)
-	// Reset hash between ranges.
+	w.annotations = append(w.annotations, a)
+	w.annotationSize = 0
+	// Reset hash between annotations.
 	w.resetHash()
 }
 
-// RangeSize returns the size of the current range.
-func (w *Writer) RangeSize() int64 {
-	return w.rangeSize
+// AnnotationSize returns the size of the current annotation.
+func (w *Writer) AnnotationSize() int64 {
```

Comment:
This name seems weird to me, I expected it to be the size of the annotation that's been passed in, but it looks like it's actually the size of the annotated bytes. How about calling this `AnnotatedBytes` instead? I think that's a lot clearer.